### PR TITLE
Github Actions CI

### PR DESCRIPTION
**Pull request by mratsim**
Given the sad state of Azure pipelines, that always timeout and take significantly more time than Travis/Jenkins, maybe Github Actions will be significantly faster.

For nim-blscurve it's about 8 times faster on very short tests.

We have 20 parallel builds on more powerful machines than all other CI, it also starts immediately.

### CODE REVIEW

**Comment by mratsim** [comment]
Seems like github target the master branch by default when we initiate Github action. This will be corrected later while a nail the pipeline.

**Comment by arnetheduck** [comment]
> Seems like github target the master branch by default when we initiate Github action. This will be corrected later while a nail the pipeline.

perhaps because the PR is opened against master?

**Comment by stefantalpalaru** [comment]
No, it's common practice for the initial CI setup to look for configuration files in the "master" branch.

The PR should still go through "devel" first, though.

**Comment by mratsim** [comment]
TODO:

- [ ] Cache Mingw to avoid draining nim-lang.org data
- [ ] Run finalization tests for some targets?
- [ ] Fix 32-bit Scrypt test #1701
- [ ] Run a 64-bit test with `BLS_FORCE_BACKEND="miracl"`

We have 20 concurrent jobs (hence 10 if we have commit+PR being build) so we can afford to use some for building end users binaries as well.

**Comment by mratsim** [comment]
The Linux 32-bit finalization is failing because https://github.com/status-im/nim-nat-traversal / miniupmp is built for x86-64

image

**Comment by mratsim** [comment]
https://github.com/status-im/nim-beacon-chain/blob/7830a42e0902615fe30386ea702192d3cedea0c9/.github/workflows/ci.yml#L10-L68

Due to several issues:
- #1723 miniupmp cross-compilation on 64-bit for 32-bit Linux
- #1701 Keystore int parsing on 32-bit
- #1724 local testnet script on Mac
- #1725 beacon node bootstrapping on Windows 32-bit

I deactivated the corresponding workflows.

In terms of speed, the unit-test complete in 15min on Linux
image

We have 20 workers.

We can test minimal finalization in about 15min, however mainnet finalization is significantly longer.

CI proposal:
- We test minimal testnet in Github Action
- We only test mainnet in Jenkins, reducing Jenkins waiting time by ~15min.
- We deactivate Azure Windows for now as their machine are too slow at the moment.
- We add back ARM64 and maybe PowerPC64 to Travis if possible.

**Comment by mratsim** [comment]
Windows-amd64 failure is related to BearSSL compilation
image

non-failing broken pipe on linux amd64 finalization (PR level)

image

failing serialization error in linux amd64 finalization (commit level)

image

**Comment by mratsim** [comment]
Windows-amd64 failure is related to BearSSL compilation
image

non-failing broken pipe on linux amd64 finalization (PR level)

image

failing serialization error in linux amd64 finalization (commit level)

image

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+        working-directory: nim-beacon-chain
+        run: |
+          make -j$ncpu ARCH_OVERRIDE=$PLATFORM DISABLE_TEST_FIXTURES_SCRIPT=1 test
+
+      - name: Run nim-beacon-chain testnet0 (minimal)
+        if: matrix.target.TEST_KIND == 'finalization-minimal'
+        shell: bash
+        working-directory: nim-beacon-chain
+        run: |
+          ./scripts/launch_local_testnet.sh --testnet 0 --nodes 4 --stop-at-epoch 5 --log-level DEBUG --disable-htop --enable-logtrace --data-dir local_testnet0_data --base-port $(( 9000 + EXECUTOR_NUMBER * 100 )) --base-rpc-port $(( 7000 + EXECUTOR_NUMBER * 100 )) --base-metrics-port $(( 8008 + EXECUTOR_NUMBER * 100 )) -- --verify-finalization --discv5:no
```
Comment (stefantalpalaru) [inline .github/workflows/ci.yml:272]: "EXECUTOR_NUMBER" is a Jenkins variable and it's used when multiple CI jobs run on the same host at the same time, without any separation. In here, you should just replace it with zero and do the math to replace those expressions.

**Pull request merged**